src/pudl/harvest/harvest.py

```diff
@@ -0,0 +1,1054 @@
+        >>> find_sample(df, key=['id'], data=['report_day'])
+        (['id'], ['report_day'])
+        >>> find_sample(df, key=['id', 'report_month'])
+        (['id', 'report_day'], [])
+        >>> find_sample(df, key=['id', 'report_month'], data=['report_day'])
+        (['id', 'report_day'], [])
+    """
+    key = list(key)
+    if data is None:
+        data = []
+    data = [col for col in data if col not in key]
+    df_key = []
+    for k in key:
+        for name in add_smaller_periods([k]):
+            if name in df.columns:
+                df_key.append(name)
+    df_data = [col for col in data if col in df.columns and col not in df_key]
+    return df_key, df_data
+
+
+def sample(
+    df: pd.DataFrame, key: Iterable[str], data: Iterable[str] = None
+) -> Optional[pd.DataFrame]:
+    """
+    Sample a dataframe.
+
+    Arguments:
+        df: Dataframe to sample.
+        key: Names of primary-key columns.
+            Takes precedence over `columns` if a name appears in both.
+        data: Names of data columns.
+
+    Returns:
+        A dataframe with all `key` columns and any matching `data` columns from `df`.
+        Values of periodic `key` columns are snapped to the start of the desired period,
+        and cast to :class:`np.datetime64`.
+        If not all `key` matched columns in `df`, nothing is returned.
+
+    Examples:
+        >>> df = pd.DataFrame({'id': [0], 'report_day': ['2020-12-31']})
+        >>> sample(df, key=['id'], data=['report_day'])
+           id  report_day
+        0   0  2020-12-31
+        >>> sample(df, key=['id', 'report_month'])
+           id report_month
+        0   0   2020-12-01
+        >>> sample(df, key=['id', 'report_quarter'])
+           id report_quarter
+        0   0     2020-10-01
+        >>> sample(df, key=['id', 'report_date']) is None
+        True
+
+        Periodic key columns are cast to :class:`np.datetime64`,
+        whether or not the period needs to be changed.
+
+        >>> sample(df, key=['report_day'])['report_day'].dtype.name
+        'datetime64[ns]'
+        >>> sample(df, key=['report_month'])['report_month'].dtype.name
+        'datetime64[ns]'
+    """
+    df_key, df_data = find_sample(df=df, key=key, data=data)
+    if len(df_key) < len(key):
+        return None
+    mapper = {old: new for old, new in zip(df_key, key) if new != old}
+    sdf = df[df_key + df_data].rename(columns=mapper)
+    for k, df_k in zip(key, df_key):
+        _, period = split_period(k)
+        if period:
+            sdf[k] = PERIODS[period](sdf[k])
+    return sdf
+
+
+# ---- Column aggregation ---- #
+
+"""
+Aggregation functions.
+
+All take a :class:`pd.Series` as input (and any optional keyword arguments).
+They may either return a single value (ideally of the same data type as the input),
+null (`np.nan`),
+or raise a :class:`ValueError` if the input does not meet requirements.
+"""
+
+
+def most_frequent(x: pd.Series) -> Any:
```

Comment:
can you add a method which is most frequent that has some threshold the occurrences of the mode in x / len of the non-null x values? Right now our main method returns values if the most frequent value is more than 70% of the reported values.

Reply:
No problem. I added `most_and_more_frequent(x, min_frequency=None)`. By default (`min_frequency=0`), it is equivalent to `most_frequent(x)`. When `min_frequency=1`, it is equivalent to `unique()`. In both cases, it is slower, so I'll leave the others in place for now.

Reply:
great!